### PR TITLE
EZP-31383: Implemented API to copy Roles

### PR DESCRIPTION
**Pull request by barw4**
| Question           | Answer
| ------------------ | ------------------
| **JIRA issue**     | [EZP-31383](https://jira.ez.no/browse/EZP-31383)
| **Improvement**| yes
| **New feature**    | yes
| **Target version** | master
| **BC breaks**      | no
| **Tests pass**     | yes
| **Doc needed**     | no

Prepared `ezpublish-kernel` for `Role` copying functionality.
Added `copyRole` method to `RoleService` API Interface with `Core` and `Event` implementations.
Created related structs and `Database Gateway` and `User\Handler` implementations.
Created `DatabaseHandler` role copying test and tests using `RoleService` API with double-check on`Policies` and `Limitations` respectively.

**TODO**:
- [x] Implement feature.
- [x] Implement tests.
- [x] Fix new code according to Coding Standards (`$ composer fix-cs`).
- [x] Ask for Code Review.

### CODE REVIEW

**Comment by barw4** [comment]
> What does it mean "Initial role copy service commit "? Please rephrase this, following our conventions, so it's meaningful. This cannot go in, in the current form. Please also describe what happens here.

The title has been changed and description extended in order to better reflect the changes.

File: eZ/Publish/Core/Repository/RoleService.php [hunk 1]
```diff
@@ -183,6 +185,73 @@ public function createRoleDraft(APIRole $role): APIRoleDraft
...
+
+        $spiRoleCopyStruct = $this->roleDomainMapper->buildPersistenceRoleCopyStruct($roleCopyStruct, $role->id);
+
+        $this->repository->beginTransaction();
+        try {
+            $spiRole = $this->userHandler->copyRole($spiRoleCopyStruct);
+            $this->repository->commit();
+        } catch (Exception $e) {
+            $this->repository->rollback();
+            throw $e;
```
Comment (alongosz) [inline eZ/Publish/Core/Repository/RoleService.php]: API should throw API exceptions only. So in the worst case, if an unknown exception is thrown, it should be wrapped with `BadStateException`. Please make sure that PhpStorm with its static analysis doesn't complain about missing throws in PhpDoc.

File: eZ/Publish/Core/Repository/RoleService.php [hunk 2]
```diff
@@ -183,6 +185,73 @@ public function createRoleDraft(APIRole $role): APIRoleDraft
...
+     * @throws \eZ\Publish\API\Repository\Exceptions\UnauthorizedException if the authenticated user is not allowed to copy a Role
+     * @throws \eZ\Publish\API\Repository\Exceptions\InvalidArgumentException if the name of the role already exists or if limitation of the same type
+     *         is repeated in the policy create struct or if limitation is not allowed on module/function
+     * @throws \eZ\Publish\API\Repository\Exceptions\LimitationValidationException if a policy limitation in the $roleCopyStruct is not valid
+     *
+     * @param \eZ\Publish\API\Repository\Values\User\Role $role
+     * @param \eZ\Publish\API\Repository\Values\User\RoleCopyStruct $roleCopyStruct
+     *
+     * @return \eZ\Publish\API\Repository\Values\User\Role
+     */
```
Comment (alongosz) [inline eZ/Publish/Core/Repository/RoleService.php]: This is redundant, the same doc is on API interface, it should be there only (no need to inheritdoc, per the latest eZ Platform Coding Style).
```suggestion
```

File: eZ/Publish/SPI/Persistence/User/RoleCopyStruct.php [hunk 3]
```diff
@@ -0,0 +1,35 @@
...
+namespace eZ\Publish\SPI\Persistence\User;
+
+use eZ\Publish\SPI\Persistence\ValueObject;
+
+class RoleCopyStruct extends ValueObject
+{
+    /**
+     * ID of user role being cloned.
+     *
+     * @var mixed
```
Comment (alongosz) [inline eZ/Publish/SPI/Persistence/User/RoleCopyStruct.php]: Please use correct type hint

File: eZ/Publish/SPI/Persistence/User/RoleCopyStruct.php [hunk 4]
```diff
@@ -0,0 +1,35 @@
...
+     * Identifier of new role.
+     *
+     * @var string
+     */
+    public $newIdentifier;
+
+    /**
+     * Contains an array of role policies.
+     *
+     * @var mixed[]
```
Comment (alongosz) [inline eZ/Publish/SPI/Persistence/User/RoleCopyStruct.php]: Please use correct type hint

File: eZ/Publish/Core/Repository/Helper/RoleDomainMapper.php [hunk 5]
```diff
@@ -200,6 +202,34 @@ public function buildPersistenceRoleCreateStruct(APIRoleCreateStruct $roleCreate
...
 
+    /**
+     * Creates SPI Role copy struct from provided API role copy struct.
+     *
+     * @param \eZ\Publish\API\Repository\Values\User\RoleCopyStruct $roleCopyStruct
+     * @param mixed $clonedId
+     *
+     * @return \eZ\Publish\SPI\Persistence\User\RoleCopyStruct
+     */
+    public function buildPersistenceRoleCopyStruct(APIRoleCopyStruct $roleCopyStruct, $clonedId)
```
Comment (alongosz) [inline eZ/Publish/Core/Repository/Helper/RoleDomainMapper.php]: Use strict types, please.

File: eZ/Publish/Core/Repository/Helper/RoleDomainMapper.php [hunk 6]
```diff
@@ -200,6 +202,34 @@ public function buildPersistenceRoleCreateStruct(APIRoleCreateStruct $roleCreate
         );
     }
 
+    /**
+     * Creates SPI Role copy struct from provided API role copy struct.
+     *
+     * @param \eZ\Publish\API\Repository\Values\User\RoleCopyStruct $roleCopyStruct
+     * @param mixed $clonedId
+     *
+     * @return \eZ\Publish\SPI\Persistence\User\RoleCopyStruct
```
Comment (alongosz) [inline eZ/Publish/Core/Repository/Helper/RoleDomainMapper.php]: Once the prototype is strict:
```suggestion
```

**Comment by barw4** [comment]
@alongosz thank you for the detailed review :) I was kinda relying too much on existing solutions I guess, I'm gonna try to fix this according to your notes.

**Comment by barw4** [comment]
closed in favor of: https://github.com/ezsystems/ezpublish-kernel/pull/3009

**Pull request closed without merging**